### DISCUSSION TITLE
Zone transfer of primary zone throws error with DNSPython due to nameserver domain

### DISCUSSION THREAD

**Discussion by PseudoResonance** [TechnitiumSoftware/DnsServer#1142] [Q&A] [answered]
I have 2 Technitium instances, one has my primary zone `example.com`, and my other has secondary zone `example.com`. My primary has a NS record `@` pointing to `dns2.internal`, with a glue address `10.0.0.10` The zone transfer/notify is setup to allow transfers/notify to listed NS records, and it works as expected, with all records from primary being propagated to secondary immediately.

Now I am using DNSPython to interact with the Technitium servers and am attempting to do a zone transfer to get the records. It works as expected for the first few records, but then I get an error `name parameter must be a subdomain of the zone origin`. Looking into it, it seems that in my zone transfer for `example.com`, Technitium is returning the record `dns2.internal` due to the NS record, but because `dns2.internal` is not a valid subdomain of `example.com`, DNSPython throws the error. I haven't had this issue with another name transfer program I am using, external-dns for Kubernetes though.

Is this out of spec on Technitium's side? I can't seem to find if it's normal for a zone transfer to include a non subdomain like that. Or is DNSPython being overly strict when it shouldn't be, and I should figure out how to get the issue resolved on that end?

Sorry if this is confusing. Thanks!

Edit: I temporarily disabled the NS record and manually added the DNS server IPs to the zone transfer/notify IP allow lists, and now it all works as expected and DNSPython can process the zone transfer. However it was nice being able to just use a single NS record to keep it more organized.

**Comment by ShreyasZare** [comment, marked as answer]
Thanks for asking. Its not recommended to set glue records for your zone's NS records. The glue records are required only when you are delegating a subdomain name to another name server. A glue record is sent as an A record in zone transfer which is why the DNSPython is giving you that error.

The proper way to do is to have a "internal" zone with A record for "dns2". The other way is how you configured the zone transfer and notify settings in Zone Options.

**Reply by PseudoResonance** [reply]
I see. I have a forwarder zone pointing to my router's DNS for internal resolution, but it didn't work and I had to add glue addresses when I used one of those domains as a name server. Is that because it's a forwarder and not authoritative or something?

**Reply by ShreyasZare** [reply]
You need to use FWD record in forwarder zone to forward to an upstream server. NS records in forwarder zone will cause it to work like a stub zone. NS record when set with a glue address will be used by the DNS server but it should be used only when required.

**Reply by PseudoResonance** [reply]
No sorry for the confusion. My setup is as follows.

---PRIMARY (example.com)---
@ NS dns2.internal

---FORWARD (internal)---
@ FWD 10.0.0.1

The NS in the primary zone points to a domain in the forward zone. The forward zone has no NS, but forwards to my router, so when I query dns2.internal, it forwards to my router and returns the A record to 10.0.0.10. But I don't think NS can point to a forwarded record, so I got errors in the logs that it couldn't find the NS. That's why I added the glue address.

Edit: It just spams `[10.0.0.10:33572] [TCP] DNS Server refused a zone transfer request since the request IP address is not allowed by the zone: example.com` in the logs almost 1000 times a second until my disk fills up if I don't have the glue address or specify notify/zone transfer IPs manually. Sometimes it also says it can't resolve `dns2.internal` for a notify, but I can't seem to get it to do that right now.

**Reply by ShreyasZare** [reply]
Thanks for the details. I tried the same zone setup here and its working well as expected. Do you have "DNSSEC Validation" option enabled for the FWD record? Because that could cause failure to resolve private domain names.

**Reply by PseudoResonance** [reply]
I was about to say no, but I do have it enabled for some reason, despite my router not supporting it... I'll try it out after disabling it.

**Reply by PseudoResonance** [reply]
It seems to be working at least more reliably now, but I still get the `DNS Server refused a zone transfer request since the request IP address is not allowed by the zone` error. I just added the IP to the zone transfer ACL list instead of allow zone name servers. Thanks for the help! I tried restarting the servers and flushing cache a few times as well, but it seemed to keep working.

**Reply by ShreyasZare** [reply]
You're welcome. Yes, the ACL option is good for such scenarios.

**Reply by PseudoResonance** [reply]
Actually, I'm pretty sure it's not supposed to be making 5,000 requests per second, right? I just looked at the dashboard, and it says there's 15 million queries in the past hour from my 2nd DNS... There is an endless barrage of this in the logs if I enable the query logging for a second.

`QNAME: example.com; QTYPE: SOA; QCLASS: IN; TSIG KeyName: tsig-keyname; TSIG Algo: hmac-sha256; TSIG Error: NoError; RCODE: NoError; ANSWER: [dns2. admin.example.com. 133 0 0 604800 0]`

Did I misconfigure something?

**Reply by ShreyasZare** [reply]
Yes, you have issue with SOA record values here. You have set refresh and retry values to 0 so its going to constantly try to refresh the zone. Update the refresh, retry, and minimum values to something practical and it will fix the issue. Note that these values are in seconds so setting refresh to 300 will mean that the zone will try to refresh every 5 mins.

**Reply by PseudoResonance** [reply]
OH! I don't know how I missed that, or why it was even set to that in the first place... Maybe I misread something, I don't know... Thank you! 3 requests per second is a lot better than 5000!

**Reply by ShreyasZare** [reply]
You're welcome. Settings a low value like 3 sec is not recommended at all since its unnecessary. Whenever primary zone is updated, it sends a NOTIFY request to secondary zones automatically. These SOA values are used as a backup just in case the NOTIFY failed or there was some network issue when trying to sync the zone.

**Reply by PseudoResonance** [reply]
I see. I'm guessing that maybe when I was having trouble with that, I tried setting it to 0 for a bit just to see, then promptly forgot about it. I just set it back to the defaults, of 900, 300, 604800, 900. Thank you very much for all the help! While I can't say the web API is particularly easy to use (all the different query parameters...) I've been really happy with Technitium since I set it up about a month ago. Plus with the standard zone transfer/dynamic updates, I can easily interface with any DNS server, which has been fantastic. Thanks for all your work!

**Reply by ShreyasZare** [reply]
You're welcome. Thanks for the compliments.
